tests/MongoDB.Bson.Tests/Jira/CSharp1559Tests.cs

```diff
@@ -0,0 +1,232 @@
+﻿/* Copyright 2020-present MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.Reflection;
+using FluentAssertions;
+using MongoDB.Bson.Serialization;
+using MongoDB.Bson.Serialization.Attributes;
+using MongoDB.Bson.TestHelpers;
+using Xunit;
+
+namespace MongoDB.Bson.Tests.Jira
+{
+    public class CSharp1559Tests
+    {
+        [Theory]
+        [InlineData(typeof(DerivedWithoutSetter_BaseWithoutSetter))]
+        [InlineData(typeof(DerivedWithoutSetter_BaseWithPrivateSetterAndWithProtectedConstructor))]
+        [InlineData(typeof(DerivedWithoutSetterAndWithBsonElement_BaseWithoutSetterAndWithProtectedConstructor))]
+        [InlineData(typeof(DerivedWithoutSetterAndWithBsonElementAndWithPrivateConstructor_BaseWithoutSetter))]
+        [InlineData(typeof(DerivedWithoutSetterAndWithoutBsonElement_AbstractBaseWithoutSetterAndWithProtectedConstructor))]
+        [InlineData(typeof(DerivedWithoutSetterAndWithoutBsonElement_AbstractBaseWithoutSetter))]
+        [InlineData(typeof(DerivedWithPrivateSetterAndWithBsonElement_BaseWithoutSetter))]
+        public void Serialization_should_work_as_expected(Type testCaseType)
+        {
+            var testCase = Activator.CreateInstance(testCaseType, 1, 2);
+            var json = testCase.ToJson();
+            var result = BsonSerializer.Deserialize(json, testCaseType);
+            var x = GetPropertyValue(result, "X");
+            var y = GetPropertyValue(result, "Y");
+            x.Should().Be(1);
+            y.Should().Be(2);
```

Comment:
Separate Arrange/Act/Assert sections with one blank line between sections?

Reply:
Would it make more sense to write asserts against the class map?

I think this testing approach is indirect... but can be OK if you prefer.

Reply:
done.
> think this testing approach is indirect... but can be OK if you prefer.

Agreed, but this way is more compact than creating serializers directly, then getting a string and trying to deserialize the value into the original object. I can update it, but this is also look good for me for the scope of this ticket

Reply:
I didn't mean creating serializers. I mean asserting against the class maps directly rather than indirectly testing, as in:

    var classMap = BsonClassMap.LookupClassMap(testCaseType);

    var constructorInfo = (ConstructorInfo)classMap.CreatorMaps.Single().MemberInfo;
    var constructorParameters = constructorInfo.GetParameters();
    constructorParameters.Length.Should().Be(2);
    constructorParameters[0].Name.Should().Be("x");
    constructorParameters[1].Name.Should().Be("y");
    var memberMaps = classMap.AllMemberMaps;
    memberMaps.Count.Should().Be(2);
    memberMaps[0].MemberName.Should().Be("X");
    memberMaps[1].MemberName.Should().Be("Y");

But what you are doing is OK too.